### PR TITLE
Add Specification to replace Problem

### PR DESCRIPTION
**Pull request by kytrinyx**
This does not remove the 'Problem' class, it inherits from it and replaces the tests
so that we can change all of the downstream dependencies without breaking anything.

See #40.

### CODE REVIEW

**Comment by Insti** [comment]
When I did this step in a dev branch, I renamed `Problem` to `Specification` and then for backwards compatibility, made a new `Problem` which is `Specification`

With the way that it's done here, `Problem` still has all the functionality, so any future changes to `Specification` look like they need to be done in the Problem class.

What is the plan for transitioning the implementation from `Problem` to `Specification` in this version?

File: README.md [hunk 1]
```diff
@@ -82,7 +82,7 @@ Bug reports and pull requests are welcome on GitHub at https://github.com/exerci
 
 See CONTRIBUTING.md for details.
 
-This is an extraction from https://github.com/exercism/x-api.
+This is extraction from https://github.com/exercism/x-api.
```
Comment (Insti) [inline README.md]: "was extracted"?

Comment (kytrinyx) [inline README.md]: I didn't mean to do this here. I rewrote x-api from scratch then extracted it to this. Let me remove this.

**Comment by kytrinyx** [comment]
> When I did this step in a dev branch, I renamed Problem to Specification and then for backwards compatibility, made a new Problem which is a Specification

This is equivalent. It doesn't matter which way you do it, it's just the same steps in a different order.

> What is the plan for transitioning the implementation from Problem to Specification in this version?

After all the downstream changes are made, then move all the functionality or delete/rename. It really doesn't matter so long as nothing breaks along the way.

**Comment by Insti** [comment]
> This is equivalent. It doesn't matter which way you do it, it's just the same steps in a different order.

I agree.

> After all the downstream changes are made, then move all the functionality or delete/rename. It really doesn't matter so long as nothing breaks along the way.

Ok. The other way gets all the big "just renaming" diffs out of the way in one step, which I think makes for easier reviews, but I guess it doesn't really matter.

**Comment by kytrinyx** [comment]
> The other way gets all the big "just renaming" diffs out of the way in one step, which I think makes for easier reviews, but I guess it doesn't really matter.

I don't have strong opinions either way on what's easier to review. I've reviewed a number of big changes that have done it either way, and I don't have a preference.

**Comment by kytrinyx** [comment]
OK, this is ready for review, if you have some time @Insti.

Each commit can be reviewed independently to make this more manageable.

**Comment by Insti** [comment]
It's too late at night for me to give this a proper review now, I can have a look at it tomorrow.

File: lib/trackler/problems.rb [hunk 2]
```diff
@@ -9,8 +9,8 @@ def initialize(root)
     end
 
     def each
-      active.each do |problem|
-        yield problem
+      active.each do |specification|
+        yield specification
       end
```
Comment (Insti) [inline lib/trackler/problems.rb]: Commit level comment: 

At this point we have a `Problems` class that only references `specifications` and never `problems`.
If we were to move the code into the `Specifications` class and reverse the dependency would that change the API at all?

**Comment by Insti** [comment]
This all looks good 👍

File: lib/trackler/problems.rb (hunk 2 above)
Comment (kytrinyx) [inline lib/trackler/problems.rb]: No that would be fine.

Comment (kytrinyx) [inline lib/trackler/problems.rb]: I've reversed the direction of the dependencies.

File: lib/trackler/problem.rb [hunk 3]
```diff
@@ -2,100 +2,10 @@
 require_relative 'metadata'
 require_relative 'description'
 require_relative 'null_track'
+require_relative 'specification'
```
Comment (Insti) [inline lib/trackler/problem.rb]: Maybe a good idea to also move these first 4 requires?

Comment (kytrinyx) [inline lib/trackler/problem.rb]: Done.

**Pull request merged**